File: ctfd.py
#!/usr/bin/env python3

import os
import re
import json
import argparse
import requests
import warnings
from pathlib import Path
from dotenv import load_dotenv
from collections import defaultdict
from urllib.parse import unquote, urljoin, urlparse

SCORE_KEYS = [
    "id",
    "name",
    "ratings",
    "category",
    "value",
    "solved_by_me",
    "attempts",
]

META_KEYS = [
    "id",
    "name",
    "description",
    "category",
    "type",
    "files",
    "hints",
    "ratings",
]

SCORE_FILE_NAME = "RESULTS.json"
META_FILE_NAME = "META.json"

def sanitize_filename(value: str, fallback: str = "unnamed") -> str:
    """
    Convert an arbitrary challenge or file name into a filesystem-safe name.
    """
    value = unquote(str(value)).strip()
    value = value.replace("\\", "_").replace("/", "_")
    value = re.sub(r'[<>:"|?*\x00-\x1f]', "_", value)
    value = re.sub(r"\s+", " ", value).strip(" .")

    return value[:180] or fallback


def get_response(url, auth):
    response = requests.get(url, headers=auth, timeout=30)

    if response.status_code != 200:
        raise ValueError(
            f"Error getting request: "
            f"{response.status_code} - {response.text}"
        )

    try:
        res = response.json()
    except:
        raise ValueError("No response as json")

    if res.get("success") is False:
        raise ValueError(f"API returned success=false: {res}")

    return res


def get_challenges(base_url, auth):
    response = get_response(
        urljoin(base_url, "challenges"),
        auth,
    )
    return response["data"]


def get_challenge(challenge, base_url, auth):
    challenge_url = urljoin(
        base_url,
        f"challenges/{challenge['id']}",
    )

    response = get_response(challenge_url, auth)
    return response["data"]


def get_file(file_url, site_url, auth):
    """
    Download a CTFd challenge file.

    Returns:
        tuple[str, bytes]: filename and file contents
    """
    full_url = urljoin(site_url, file_url)

    response = requests.get(
        full_url,
        headers=auth,
        timeout=60,
    )

    if response.status_code != 200:
        raise ValueError(
            f"Error downloading {full_url}: "
            f"{response.status_code} - {response.text}"
        )

    # Ignore the query string, including the signed token.
    filename = Path(
        unquote(urlparse(full_url).path)
    ).name

    filename = sanitize_filename(filename, "downloaded_file")

    return filename, response.content


def save_challenge(challenge, location, site_url, auth):
    global META_KEYS
    meta = {
        key: value
        for key, value in challenge.items()
        if key in META_KEYS
    }

    meta["file_paths"] = meta.get("files", [])
    meta["files"] = []

    category = sanitize_filename(
        challenge.get("category", "UNKNOWN")
    ).replace(" ", "_")

    challenge_name = sanitize_filename(
        challenge.get("name", "UNKNOWN")
    ).replace(" ", "_")

    parent = Path(location) / category
    challenge_dir = parent / challenge_name

    challenge_dir.mkdir(parents=True, exist_ok=True)

    for file_url in challenge.get("files", []):
        filename, file_contents = get_file(
            file_url,
            site_url,
            auth,
        )

        with open(challenge_dir / filename, "wb") as f:
            f.write(file_contents)

        meta["files"].append(filename)

    with open(
        challenge_dir / META_FILE_NAME,
        "w",
        encoding="utf-8",
    ) as f:
        json.dump(meta, f, indent=2)

def format_score_output(challenges):
    grouped = {
        "SOLVED": defaultdict(list),
        "UNSOLVED": defaultdict(list),
    }

    for challenge in challenges:
        section = "SOLVED" if challenge.get("solved_by_me") else "UNSOLVED"
        rating = challenge.get("ratings", "UNRATED")

        grouped[section][rating].append(challenge)

    for section in ("SOLVED", "UNSOLVED"):
        print(f"{section}:")

        # Sort rating groups by the highest-point challenge in each group.
        ratings = sorted(
            grouped[section],
            key=lambda rating: max(
                challenge.get("value", 0)
                for challenge in grouped[section][rating]
            ),
            reverse=True,
        )

        for rating in ratings:
            print(f"    {rating}:")

            challenges_for_rating = sorted(
                grouped[section][rating],
                key=lambda challenge: challenge.get("value", 0),
                reverse=True,
            )

            for challenge in challenges_for_rating:
                name = challenge.get("name", "UNKNOWN")
                category = challenge.get("category", "UNKNOWN")
                value = challenge.get("value", 0)
                attempts = challenge.get("attempts", 0)

                print(
                    f"        {name}: {category} "
                    f"({value} pts, {attempts} tries)"
                )

        print()

def pull_event(site_url, location, auth):
    base_url = urljoin(site_url, "api/v1/")

    challenges = get_challenges(base_url, auth)
    failed = []
    for challenge in challenges:
        try:
            full_challenge = get_challenge(
                challenge,
                base_url,
                auth,
            )

            save_challenge(
                full_challenge,
                location,
                site_url,
                auth,
            )

            print(
                f"Saved: {full_challenge.get('name', 'UNKNOWN')}"
            )

        except Exception as e:
            failed.append(
                {
                    "id": challenge.get("id"),
                    "name": challenge.get("name"),
                    "err": str(e),
                }
            )

    if failed:
        print(f"Failed:\n{json.dumps(failed, indent=2)}")

def score_event(site_url, location, auth):
    global SCORE_KEYS
    base_url = urljoin(site_url, "api/v1/")


    challenges = get_challenges(base_url, auth)
    results = []
    failed = []
    for challenge in challenges:
        try:
            full_challenge = get_challenge(
                challenge,
                base_url,
                auth,
            )

            results.append({
                key: value
                for key, value in full_challenge.items()
                if key in SCORE_KEYS
            })

            print(
                f"Scored: {full_challenge.get('name', 'UNKNOWN')}"
            )

        except Exception as e:
            failed.append(
                {
                    "id": challenge.get("id"),
                    "name": challenge.get("name"),
                    "err": str(e),
                }
            )

    if failed:
        print(f"Failed:\n{json.dumps(failed, indent=2)}")

    with open(Path(location) / SCORE_FILE_NAME, "w") as f:
        json.dump(results, f, indent=2)
    
    format_score_output(results)

def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog="ctfd.py",
        description="Pull and score challenges from a CTFd instance.",
    )

    subparsers = parser.add_subparsers(
        dest="command",
        required=True,
    )

    # Shared arguments for both pull and score
    def add_common_arguments(subparser: argparse.ArgumentParser) -> None:
        subparser.add_argument(
            "site_url",
            help="Base URL of the CTFd site",
        )
        subparser.add_argument(
            "location",
            help="Local challenge directory",
        )
        subparser.add_argument(
            "--session",
            dest="session_cookie",
            help="CTFd session cookie",
        )
        subparser.add_argument(
            "--auth",
            dest="auth_token",
            help="CTFd API authentication token",
        )

    pull_parser = subparsers.add_parser(
        "pull",
        help="Download challenges from a CTFd site",
    )
    add_common_arguments(pull_parser)

    score_parser = subparsers.add_parser(
        "score",
        help="Score event results",
    )
    add_common_arguments(score_parser)

    return parser

def main():
    parser = build_parser()

    args = parser.parse_args()

    load_dotenv()

    auth = dict()
    if args.auth_token:
        auth["Authorization"]= f"Token {args.auth_token}"
    elif os.getenv("CTFd_TOKEN"):
        auth["Authorization"]= f"Token {os.getenv('CTFd_TOKEN')}"
    
    if args.session_cookie:
        auth["Cookie"] = args.session_cookie

    if not auth:
        warnings.warn("No Authorization Provided")

    if args.command == "pull":
        pull_event(args.site_url, args.location, auth)
    elif args.command == "score":
        score_event(args.site_url, args.location, auth)

if __name__ == "__main__":
    main()

File: ctfd.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"github.com/vbauerster/mpb/v8"
	"github.com/vbauerster/mpb/v8/decor"
	"golang.org/x/sync/errgroup"
	"io"
	"net/http"
	"net/url"
	"os"
	"path"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
)

type Command string

const (
	CommandPull  Command = "pull"
	CommandScore Command = "score"
	CommandStore Command = "store"
)

type CLIOptions struct {
	Command  Command
	Location string

	BaseURL    string
	BaseURLSet bool

	Cookie    string
	CookieSet bool

	GroupLimit    int
	GroupLimitSet bool
}

const (
	MetadataFilename   = "META.json"
	ScoresFilename     = "SCORE.json"
	DefaultGroupLimit  = 5
	DefaultFolderPerms = 0o755
	DefaultFilePerms   = 0o644
)

type CTFdEvent struct {
	BaseURL                          string
	Headers                          map[string]string
	SuccessfullyDownloadedChallenges map[int]ChallengeInfo
	Location                         string
	Progress                         *mpb.Progress
	GroupLimit                       int
}

type CTFdEventMetadata struct {
	BaseURL                          string            `json:"base_url"`
	Headers                          map[string]string `json:"headers"`
	SuccessfullyDownloadedChallenges []ChallengeInfo   `json:"known_challenges"`
	Location                         string            `json:"location"`
}

func JoinUrl(baseUrl string, parts ...string) (string, error) {
	return url.JoinPath(baseUrl, parts...)
}

func ResolveURL(baseUrl string, reference string) (string, error) {
	base, err := url.Parse(baseUrl)
	if err != nil {
		return "", fmt.Errorf("parse base URL: %w", err)
	}

	ref, err := url.Parse(reference)
	if err != nil {
		return "", fmt.Errorf("parse URL reference %q: %w", reference, err)
	}

	return base.ResolveReference(ref).String(), nil
}

func Get(url string, headers map[string]string) ([]byte, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}

	for key, value := range headers {
		req.Header.Set(key, value)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	return body, nil
}

type Hint struct {
	Id      int    `json:"id"`
	Cost    int    `json:"cost"`
	Title   string `json:"title"`
	Content string `json:"content"`
}

type DownloadedFile struct {
	Name string
	Data []byte
}

type Challenge struct {
	Id         int    `json:"id"`
	Name       string `json:"name"`
	Value      int    `json:"value"`
	Category   string `json:"category"`
	SolvedByMe bool   `json:"solved_by_me"`
	Attempts   int    `json:"attempts"`
	Ratings    any    `json:"ratings"`

	Description string            `json:"description"`
	Type        string            `json:"type"`
	Files       []*DownloadedFile `json:"-"`
	FilePaths   []string          `json:"files"`
	Hints       []Hint            `json:"hints"`
	MaxAttempts int               `json:"max_attempts"`
}

type ChallengeMetadata struct {
	Id          int      `json:"id"`
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Category    string   `json:"category"`
	Type        string   `json:"type"`
	Files       []string `json:"files"`
	Hints       []Hint   `json:"hints"`
	Ratings     any      `json:"ratings"`
}

type APIResponse struct {
	Success bool `json:"success"`
	Data    any  `json:"data"`
}

type ChallengeInfo struct {
	Id   int    `json:"id"`
	Name string `json:"name"`
}

type Challenges []ChallengeInfo
type FinalScore struct {
	Score      int `json:"score"`
	Total      int `json:"total"`
	Solved     int `json:"solved"`
	Unsolved   int `json:"unsolved"`
	Failed     int `json:"failed"`
	Challenges int `json:"challenge_count"`
}

type ChallengeScores struct {
	Scores      []ChallengeScore `json:"scores"`
	FinalResult FinalScore       `json:"results"`
}

func (ctfd *CTFdEvent) GetChallenges() (*Challenges, error) {
	endpoint, err := JoinUrl(ctfd.BaseURL, "/api/v1/challenges")
	if err != nil {
		return nil, err
	}

	resp, err := Get(endpoint, ctfd.Headers)
	if err != nil {
		return nil, err
	}

	var result APIResponse
	if err := json.NewDecoder(bytes.NewReader(resp)).Decode(&result); err != nil {
		return nil, err
	}

	if !result.Success {
		return nil, fmt.Errorf("API call returned success=false")
	}

	items, ok := result.Data.([]any)
	if !ok {
		return nil, fmt.Errorf(
			"expected API data to be an array, got %T",
			result.Data,
		)
	}

	data := make([]map[string]any, 0, len(items))

	for _, item := range items {
		challenge, ok := item.(map[string]any)
		if !ok {
			return nil, fmt.Errorf(
				"expected challenge to be an object, got %T",
				item,
			)
		}

		data = append(data, challenge)
	}

	raw, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	var challenges Challenges
	if err := json.Unmarshal(raw, &challenges); err != nil {
		return nil, err
	}

	return &challenges, nil
}

func (ctfd *CTFdEvent) GetChallenge(id int) (*Challenge, error) {
	endpoint, err := JoinUrl(ctfd.BaseURL, "/api/v1/challenges", strconv.Itoa(id))
	if err != nil {
		return nil, err
	}

	resp, err := Get(endpoint, ctfd.Headers)
	if err != nil {
		return nil, err
	}

	var result APIResponse
	if err := json.NewDecoder(bytes.NewReader(resp)).Decode(&result); err != nil {
		return nil, err
	}

	if !result.Success {
		return nil, fmt.Errorf("API Call success false")
	}

	data, ok := result.Data.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("API Returned Weird Data")
	}

	raw, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	challenge := Challenge{}
	if err := json.Unmarshal(raw, &challenge); err != nil {
		return nil, err
	}

	return &challenge, nil
}

func (ctfd *CTFdEvent) DownloadFiles(filePaths []string) ([]*DownloadedFile, error) {
	files := []*DownloadedFile{}

	for _, filePath := range filePaths {
		endpoint, err := ResolveURL(ctfd.BaseURL, filePath)
		if err != nil {
			return nil, err
		}

		resp, err := Get(endpoint, ctfd.Headers)
		if err != nil {
			return nil, err
		}

		parsed, err := url.Parse(filePath)
		if err != nil {
			panic(err)
		}

		filename := path.Base(parsed.Path)

		file := DownloadedFile{
			Name: filename,
			Data: resp,
		}

		files = append(files, &file)
	}

	return files, nil
}

func SanitizeFilename(value string) string {
	value = strings.TrimSpace(value)

	replacer := strings.NewReplacer(
		"/", "",
		"\\", "",
		":", "",
		"*", "",
		"?", "",
		`"`, "",
		"<", "",
		">", "",
		"|", "",
		"'", "",
		"\"", "",
		" ", "_",
	)

	value = replacer.Replace(value)
	value = strings.Trim(value, ". ")

	if value == "" {
		return "unnamed"
	}

	return value
}

func (challenge *Challenge) WriteChallenge(location string) error {
	challengeDir := filepath.Join(
		location,
		SanitizeFilename(challenge.Category),
		SanitizeFilename(challenge.Name),
	)

	if err := os.MkdirAll(challengeDir, DefaultFolderPerms); err != nil {
		return err
	}

	fileNames := make([]string, 0, len(challenge.Files))

	for _, file := range challenge.Files {
		if file == nil {
			continue
		}

		filename := SanitizeFilename(file.Name)
		filePath := filepath.Join(challengeDir, filename)

		if err := os.WriteFile(filePath, file.Data, DefaultFilePerms); err != nil {
			return err
		}

		fileNames = append(fileNames, filename)
	}

	metadata := ChallengeMetadata{
		Id:          challenge.Id,
		Name:        challenge.Name,
		Description: challenge.Description,
		Category:    challenge.Category,
		Type:        challenge.Type,
		Files:       fileNames,
		Hints:       challenge.Hints,
		Ratings:     challenge.Ratings,
	}

	metadataJSON, err := json.MarshalIndent(metadata, "", "  ")
	if err != nil {
		return err
	}

	metadataPath := filepath.Join(challengeDir, MetadataFilename)

	metadataJSON = append(metadataJSON, '\n')

	if err := os.WriteFile(metadataPath, metadataJSON, DefaultFilePerms); err != nil {
		return err
	}

	return nil
}

func (ctfd *CTFdEvent) SaveChallenge(chal ChallengeInfo) error {
	challenge, err := ctfd.GetChallenge(chal.Id)
	if err != nil {
		return fmt.Errorf("%d: %w", chal.Id, err)
	}

	if len(challenge.FilePaths) > 0 {
		files, err := ctfd.DownloadFiles(challenge.FilePaths)
		if err != nil {
			return fmt.Errorf("%d: %w", chal.Id, err)
		}
		challenge.Files = files
	}

	err = challenge.WriteChallenge(ctfd.Location)
	if err != nil {
		return fmt.Errorf("%d: %w", chal.Id, err)
	}
	return nil
}

type ChallengeScore struct {
	Id          int    `json:"id"`
	Name        string `json:"name"`
	Value       int    `json:"value"`
	Category    string `json:"category"`
	SolvedByMe  bool   `json:"solved_by_me"`
	Attempts    int    `json:"attempts"`
	Ratings     any    `json:"ratings"`
	MaxAttempts int    `json:"max_attempts"`
}

func (ctfd *CTFdEvent) ScoreChallenge(chal ChallengeInfo) (*ChallengeScore, error) {
	challenge, err := ctfd.GetChallenge(chal.Id)
	if err != nil {
		return nil, fmt.Errorf("%d: %v", chal.Id, err)
	}

	score := ChallengeScore{
		Id:          challenge.Id,
		Name:        challenge.Name,
		Value:       challenge.Value,
		Category:    challenge.Category,
		SolvedByMe:  challenge.SolvedByMe,
		Attempts:    challenge.Attempts,
		Ratings:     challenge.Ratings,
		MaxAttempts: challenge.MaxAttempts,
	}

	return &score, nil
}

func (ctfd *CTFdEvent) AlreadySaved(challenge ChallengeInfo) bool {
	chal, ok := ctfd.SuccessfullyDownloadedChallenges[challenge.Id]
	if !ok {
		return false
	}
	if chal.Name == challenge.Name {
		return true
	}

	return false
}

func (ctfd *CTFdEvent) PullEvent() {
	challengeList, err := ctfd.GetChallenges()
	if err != nil {
		panic(err)
	}

	bar := ctfd.Progress.AddBar(
		int64(len(*challengeList)),
		mpb.PrependDecorators(
			decor.Name("Loading "),
			decor.CountersNoUnit("%d / %d"),
		),
		mpb.AppendDecorators(
			decor.Percentage(),
			decor.Elapsed(decor.ET_STYLE_GO),
		),
	)

	mutex := sync.Mutex{}
	fails := []ChallengeInfo{}
	var group errgroup.Group

	group.SetLimit(ctfd.GroupLimit)

	for _, chal := range *challengeList {
		if ctfd.AlreadySaved(chal) {
			bar.Increment()
			continue
		}

		group.Go(func() error {
			defer bar.Increment()
			err := ctfd.SaveChallenge(chal)
			if err != nil {
				mutex.Lock()
				fails = append(fails, chal)
				mutex.Unlock()
			} else {
				mutex.Lock()
				ctfd.SuccessfullyDownloadedChallenges[chal.Id] = chal
				mutex.Unlock()
			}

			return nil
		})
	}

	group.Wait()

	if len(fails) > 0 {
		fmt.Println("\nFailed:")
		for _, chal := range fails {
			fmt.Printf("\"%s\" -- id: %d\n", chal.Name, chal.Id)
		}
	}
}

func (scores ChallengeScores) WriteScores(location string) error {
	if err := os.MkdirAll(location, DefaultFolderPerms); err != nil {
		return err
	}

	data, err := json.MarshalIndent(scores, "", "  ")
	if err != nil {
		return err
	}

	data = append(data, '\n')

	metadataPath := filepath.Join(location, ScoresFilename)

	if err := os.WriteFile(metadataPath, data, DefaultFilePerms); err != nil {
		return err
	}

	return nil
}

func (ctfd *CTFdEvent) ScoreEvent() {
	challengeList, err := ctfd.GetChallenges()
	if err != nil {
		panic(err)
	}

	bar := ctfd.Progress.AddBar(
		int64(len(*challengeList)),
		mpb.PrependDecorators(
			decor.Name("Scoring "),
			decor.CountersNoUnit("%d / %d"),
		),
		mpb.AppendDecorators(
			decor.Percentage(),
			decor.Elapsed(decor.ET_STYLE_GO),
		),
	)

	var group errgroup.Group

	group.SetLimit(ctfd.GroupLimit)

	mutex := sync.Mutex{}
	scores := ChallengeScores{}
	scores.FinalResult = FinalScore{}

	for _, chal := range *challengeList {
		group.Go(func() error {
			defer bar.Increment()
			score, err := ctfd.ScoreChallenge(chal)
			if err != nil {
				return fmt.Errorf("Could not score \"%s\"", chal.Name)
			}

			mutex.Lock()
			scores.FinalResult.Challenges += 1
			scores.Scores = append(scores.Scores, *score)
			scores.FinalResult.Total += score.Value
			if score.SolvedByMe {
				scores.FinalResult.Solved += 1
				scores.FinalResult.Score += score.Value
			} else if score.MaxAttempts > 0 && score.Attempts >= score.MaxAttempts {
				scores.FinalResult.Failed += 1
			} else {
				scores.FinalResult.Unsolved += 1
			}
			mutex.Unlock()
			return nil
		})
	}

	if err := group.Wait(); err != nil {
		panic(err)
	}

	if err = scores.WriteScores(ctfd.Location); err != nil {
		panic(err)
	}

	fmt.Printf("\nSolved: %d / %d\nUnsolved: %d / %d\nFailed: %d / %d\nScore: %d / %d\n",
		scores.FinalResult.Solved,
		scores.FinalResult.Challenges,
		scores.FinalResult.Unsolved,
		scores.FinalResult.Challenges,
		scores.FinalResult.Failed,
		scores.FinalResult.Challenges,
		scores.FinalResult.Score,
		scores.FinalResult.Total)
}

func (ctfd *CTFdEvent) GetSuccessfullyDownloadedChallengeList() []ChallengeInfo {
	result := []ChallengeInfo{}
	for _, value := range ctfd.SuccessfullyDownloadedChallenges {
		result = append(result, value)
	}
	return result
}

func (ctfd *CTFdEvent) WriteMetadata() error {
	if err := os.MkdirAll(ctfd.Location, DefaultFolderPerms); err != nil {
		return fmt.Errorf("create event directory %q: %w", ctfd.Location, err)
	}

	metadata := CTFdEventMetadata{
		BaseURL:                          ctfd.BaseURL,
		Headers:                          ctfd.Headers,
		SuccessfullyDownloadedChallenges: ctfd.GetSuccessfullyDownloadedChallengeList(),
		Location:                         ctfd.Location,
	}

	data, err := json.MarshalIndent(metadata, "", "  ")
	if err != nil {
		return fmt.Errorf("encode event metadata: %w", err)
	}

	data = append(data, '\n')

	metadataPath := filepath.Join(metadata.Location, MetadataFilename)

	if err := os.WriteFile(metadataPath, data, DefaultFilePerms); err != nil {
		return fmt.Errorf("write event metadata %q: %w", metadataPath, err)
	}

	return nil
}

func ReadEventMetadata(location string) (*CTFdEvent, error) {
	absoluteLocation, err := filepath.Abs(location)
	if err != nil {
		return nil, fmt.Errorf("resolve event location %q: %w", location, err)
	}

	metadataPath := filepath.Join(
		absoluteLocation,
		MetadataFilename,
	)

	data, err := os.ReadFile(metadataPath)
	if err != nil {
		return nil, fmt.Errorf("read event metadata %q: %w", metadataPath, err)
	}

	var metadata CTFdEventMetadata

	if err := json.Unmarshal(data, &metadata); err != nil {
		return nil, fmt.Errorf("decode event metadata %q: %w", metadataPath, err)
	}

	event := CTFdEvent{
		BaseURL:                          metadata.BaseURL,
		Headers:                          metadata.Headers,
		SuccessfullyDownloadedChallenges: make(map[int]ChallengeInfo),
		Location:                         metadata.Location,
	}

	for _, chal := range metadata.SuccessfullyDownloadedChallenges {
		event.SuccessfullyDownloadedChallenges[chal.Id] = chal
	}

	return &event, nil
}

func ParseCLI() (*CLIOptions, error) {
	if len(os.Args) < 3 {
		return nil, fmt.Errorf(
			"usage: %s <score|pull|store> <location> [--base URL] [--cookie COOKIE] [--group_limit N]",
			os.Args[0],
		)
	}

	command := Command(os.Args[1])
	location := os.Args[2]

	switch command {
	case CommandPull, CommandScore, CommandStore:
	default:
		return nil, fmt.Errorf(
			"invalid command %q: expected score, pull, or store",
			command,
		)
	}

	flags := flag.NewFlagSet(string(command), flag.ContinueOnError)

	baseURL := flags.String(
		"base",
		"",
		"base URL of the CTFd instance",
	)

	cookie := flags.String(
		"cookie",
		"",
		"CTFd session cookie",
	)

	groupLimit := flags.Int(
		"group_limit",
		DefaultGroupLimit,
		"maximum number of concurrent requests",
	)

	if err := flags.Parse(os.Args[3:]); err != nil {
		return nil, err
	}

	if flags.NArg() > 0 {
		return nil, fmt.Errorf(
			"unexpected positional arguments: %v",
			flags.Args(),
		)
	}

	// Record which flags were explicitly passed.
	present := make(map[string]bool)

	flags.Visit(func(f *flag.Flag) {
		present[f.Name] = true
	})

	if present["group_limit"] && *groupLimit < 1 {
		return nil, fmt.Errorf(
			"--group_limit must be at least 1, got %d",
			*groupLimit,
		)
	}

	absoluteLocation, err := filepath.Abs(location)
	if err != nil {
		return nil, fmt.Errorf(
			"resolve directory %q: %w",
			location,
			err,
		)
	}

	return &CLIOptions{
		Command:  command,
		Location: absoluteLocation,

		BaseURL:    strings.TrimRight(*baseURL, "/"),
		BaseURLSet: present["base"],

		Cookie:    *cookie,
		CookieSet: present["cookie"],

		GroupLimit:    *groupLimit,
		GroupLimitSet: present["group_limit"],
	}, nil
}

func main() {
	options, err := ParseCLI()
	if err != nil {
		fmt.Fprintln(os.Stderr, "error:", err)
		os.Exit(2)
	}

	// First try loading the existing event from the positional directory.
	event, err := ReadEventMetadata(options.Location)
	if err != nil {
		// No usable metadata, so start with defaults.
		event = &CTFdEvent{
			BaseURL:                          "",
			Headers:                          make(map[string]string),
			Location:                         options.Location,
			GroupLimit:                       DefaultGroupLimit,
			SuccessfullyDownloadedChallenges: make(map[int]ChallengeInfo),
		}
	}

	// The positional location always wins.
	event.Location = options.Location

	if options.BaseURLSet {
		event.BaseURL = options.BaseURL
	}

	if options.CookieSet {
		if event.Headers == nil {
			event.Headers = make(map[string]string)
		}

		event.Headers["Cookie"] = options.Cookie
	}

	event.GroupLimit = options.GroupLimit
	if event.GroupLimit < 1 {
		event.GroupLimit = DefaultGroupLimit
	}

	if event.BaseURL == "" {
		fmt.Fprintln(os.Stderr, "No Base URL")
		os.Exit(1)
	}

	event.Progress = mpb.New()
	defer event.Progress.Wait()

	switch options.Command {
	case CommandPull:
		event.PullEvent()

		if err := event.WriteMetadata(); err != nil {
			fmt.Fprintln(os.Stderr, "error:", err)
			os.Exit(1)
		}

	case CommandScore:
		event.ScoreEvent()
	case CommandStore:
		if err := event.WriteMetadata(); err != nil {
			fmt.Fprintln(os.Stderr, "error:", err)
			os.Exit(1)
		}
	}
}
